File: src/main/java/net/somrpg/swordofmagic7/translater/Japanizer.java
package net.somrpg.swordofmagic7.translater;

import java.util.HashMap;
import java.util.Map;

/**
 * ローマ字表記を漢字変換して返すユーティリティ
 * @author ucchy
 */
public class Japanizer {

    public Japanizer() {
        throw new UnsupportedOperationException("Utility class");
    }

    private static final String REGEX_URL = "https?://[\\w/:%#$&?()~.=+\\-]+";

    /**
     * メッセージの日本語化をする
     * @param org メッセージ
     * @param type 日本語化のタイプ
     * @param dictionary 辞書
     * @return 日本語化したメッセージ
     */
    public static String japanize(String org, JapanizeType type,
                                  Map<String, String> dictionary) {

        // 変換不要なら空文字列を返す
        if ( type == JapanizeType.NONE || !isNeedToJapanize(org) ) {
            return "";
        }

        // URL削除
        String deletedURL = org.replaceAll(REGEX_URL, " ");

        // キーワードをロック
        HashMap<String, String> keywordMap = new HashMap<>();
        int index = 0;
        String keywordLocked = deletedURL;
        for ( String dickey : dictionary.keySet() ) {
            if ( keywordLocked.contains(dickey) ) {
                index++;
                String key = "＜" + makeMultibytesDigit(index) + "＞";
                keywordLocked = keywordLocked.replace(dickey, key);
                keywordMap.put(key, dictionary.get(dickey));
            }
        }

        // カナ変換
        String japanized = YukiKanaConverter.conv(keywordLocked);

        // IME変換
        if ( type == JapanizeType.GOOGLE_IME ) {
            japanized = IMEConverter.convByGoogleIME(japanized);
        }

        // キーワードのアンロック
        for ( String key : keywordMap.keySet() ) {
            japanized = japanized.replace(key, keywordMap.get(key));
        }

        // 返す
        return japanized.trim();
    }

    /**
     * 日本語化が必要かどうかを判定する
     * @param org メッセージ
     * @return 日本語化が必要ならtrue
     */
   public static boolean isNeedToJapanize(String org) {
        return ( org.getBytes().length == org.length()
                && !org.matches("[ \\uFF61-\\uFF9F]+") );
    }

    /**
     * 数値を、全角文字の文字列に変換して返す
     * @param digit 文字
     * @return 全角文字
     */
    private static String makeMultibytesDigit(int digit) {

        String half = Integer.toString(digit);
        StringBuilder result = new StringBuilder();
        for ( int index=0; index < half.length(); index++ ) {
            switch ( half.charAt(index) ) {
                case '0' : result.append("０"); break;
                case '1' : result.append("１"); break;
                case '2' : result.append("２"); break;
                case '3' : result.append("３"); break;
                case '4' : result.append("４"); break;
                case '5' : result.append("５"); break;
                case '6' : result.append("６"); break;
                case '7' : result.append("７"); break;
                case '8' : result.append("８"); break;
                case '9' : result.append("９"); break;
            }
        }
        return result.toString();
    }
}

File: src/main/kotlin/net/somrpg/swordofmagic7/translater/Japanizer.kt
/*
 * @author     ucchy
 * @license    LGPLv3
 * @copyright  Copyright ucchy 2020
 *
 * Porting from Java to Kotlin
 */
package net.somrpg.swordofmagic7.translater

/**
 * ローマ字表記を漢字変換して返すユーティリティ
 */
object Japanizer {
    private val REGEX_URL = Regex("https?://[\\w/:%#$&?()~.=+\\-]+")

    /**
     * メッセージの日本語化をする
     */
    @JvmStatic
    fun japanize(
        org: String,
        dictionary: Map<String, String>,
    ): String {
        val deletedURL = org.replace(REGEX_URL, " ")

        val keywordMap = HashMap<String, String>()
        var index = 0
        var keywordLocked = deletedURL
        for (dickey in dictionary.keys) {
            if (keywordLocked.contains(dickey)) {
                index++
                val key = "＜${makeMultibytesDigit(index)}＞"
                keywordLocked = keywordLocked.replace(dickey, key)
                keywordMap[key] = dictionary[dickey]!!
            }
        }

        var japanized = YukiKanaConverter.conv(keywordLocked)
        japanized = IMEConverter.convByGoogleIME(japanized)

        for ((key, value) in keywordMap) {
            japanized = japanized.replace(key, value)
        }

        val result = japanized.trim()
        return result
    }

    /**
     * 日本語化が必要かどうかを判定する
     */
    @JvmStatic
    fun isNeedToJapanize(org: String): Boolean {
        val byteLength = org.toByteArray().size
        val charLength = org.length
        val isHalfWidthAscii = byteLength == charLength
        val isHalfWidthKatakanaOnly = org.matches(Regex("[ \\uFF61-\\uFF9F]+"))
        val result = isHalfWidthAscii && !isHalfWidthKatakanaOnly
        return result
    }

    private fun makeMultibytesDigit(digit: Int): String {
        val half = digit.toString()
        val result = StringBuilder()
        for (c in half) {
            result.append(
                when (c) {
                    '0' -> "０"
                    '1' -> "１"
                    '2' -> "２"
                    '3' -> "３"
                    '4' -> "４"
                    '5' -> "５"
                    '6' -> "６"
                    '7' -> "７"
                    '8' -> "８"
                    '9' -> "９"
                    else -> ""
                },
            )
        }
        return result.toString()
    }
}
